fedimint-core/src/encoding/mod.rs

```diff
@@ -493,19 +493,76 @@ where
     }
 }
 
+impl<T> Encodable for VecDeque<T>
+where
+    T: Encodable + 'static,
+{
+    fn consensus_encode<W: std::io::Write>(&self, writer: &mut W) -> Result<usize, Error> {
+        let mut len = (self.len() as u64).consensus_encode(writer)?;
+        for i in self {
+            len += i.consensus_encode(writer)?;
+        }
+        Ok(len)
+    }
+}
+
 #[test]
 fn vec_decode_sanity() {
     let buf = [
         0xffu8, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
     ];
 
     // On malicious large len, return an error instead of panicking.
-    // Note: This was supposed to expose the panic, but I was not able to trigger it
-    // for some reason.
     assert!(Vec::<u8>::consensus_decode(&mut buf.as_slice(), &Default::default()).is_err());
     assert!(Vec::<u16>::consensus_decode(&mut buf.as_slice(), &Default::default()).is_err());
 }
 
+impl<T> Decodable for VecDeque<T>
+where
+    T: Decodable + 'static,
+{
+    fn consensus_decode_from_finite_reader<D: std::io::Read>(
+        d: &mut D,
+        modules: &ModuleDecoderRegistry,
+    ) -> Result<Self, DecodeError> {
+        let len = u64::consensus_decode_from_finite_reader(d, modules)?;
+
+        // `collect` under the hood uses `FromIter::from_iter`, which can potentially be
+        // backed by code like:
+        // <https://github.com/rust-lang/rust/blob/fe03b46ee4688a99d7155b4f9dcd875b6903952d/library/alloc/src/vec/spec_from_iter_nested.rs#L31>
+        // This can take `size_hint` from input iterator and pre-allocate memory
+        // upfront with `VecDeque::with_capacity`. Because of that untrusted `len`
+        // should not be used directly.
+        let cap_len = cmp::min(8_000 / mem::size_of::<T>() as u64, len);
+
+        // Up to a cap, use the (potentially specialized for better perf in stdlib)
+        // `from_iter`.
+        let mut v: VecDeque<_> = (0..cap_len)
+            .map(|_| T::consensus_decode_from_finite_reader(d, modules))
+            .collect::<Result<VecDeque<_>, DecodeError>>()?;
+
+        // Add any excess manually avoiding any surprises.
+        while (v.len() as u64) < len {
+            v.push_back(T::consensus_decode_from_finite_reader(d, modules)?);
+        }
+
+        assert_eq!(v.len() as u64, len);
+
+        Ok(v)
```

Comment:
can we just decode as Vec and do `VecDeque::from`?

Reply:
You are absolutely correct. :brain: